File: java/leetcode_grind/Day712.java
package leetcode_grind;

import java.util.*;

public class Day712 {
    // https://leetcode.com/problems/maximum-number-of-moves-in-a-grid/description/?envType=daily-question&envId=2024-10-29
    static class Solution1 {
        int[] dirs = { -1, 0, 1 };

        public int maxMoves(int[][] grid) {
            int M = grid.length, N = grid[0].length;
            Queue<int[]> q = new LinkedList<>();
            boolean[][] vis = new boolean[M][N];

            for (int i = 0; i < M; i++) {
                vis[i][0] = true;
                q.offer(new int[] { i, 0, 0 });
            }

            int maxMoves = 0;
            while (!q.isEmpty()) {
                int sz = q.size();
                while (sz-- > 0) {
                    int[] v = q.poll();
                    int row = v[0], col = v[1], count = v[2];
                    maxMoves = Math.max(maxMoves, count);

                    for (int dir : dirs) {
                        int newRow = row + dir, newCol = col + 1;

                        if (newRow >= 0 && newCol >= 0 && newRow < M && newCol < N && !vis[newRow][newCol]
                                && grid[row][col] < grid[newRow][newCol]) {
                            vis[newRow][newCol] = true;
                            q.offer(new int[] { newRow, newCol, count + 1 });
                        }
                    }
                }
            }
            return maxMoves;
        }
    }

    static class Solution2 {
        final int[] dirs = { -1, 0, 1 };

        int DFS(int row, int col, int[][] grid, int[][] dp) {
            int M = grid.length, N = grid[0].length;
            if (dp[row][col] != -1) {
                return dp[row][col];
            }
            int maxMoves = 0;
            for (int dir : dirs) {
                int newRow = row + dir, newCol = col + 1;

                if (newRow >= 0 &&
                        newCol >= 0 &&
                        newRow < M &&
                        newCol < N &&
                        grid[row][col] < grid[newRow][newCol]) {
                    maxMoves = Math.max(
                            maxMoves,
                            1 + DFS(newRow, newCol, grid, dp));
                }
            }

            dp[row][col] = maxMoves;
            return maxMoves;
        }

        public int maxMoves(int[][] grid) {
            int M = grid.length, N = grid[0].length;
            int[][] dp = new int[M][N];
            for (int i = 0; i < M; i++) {
                Arrays.fill(dp[i], -1);
            }
            int maxMoves = 0;
            for (int i = 0; i < M; i++) {
                int movesRequired = DFS(i, 0, grid, dp);
                maxMoves = Math.max(maxMoves, movesRequired);
            }
            return maxMoves;
        }
    }

    // https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description/
    static class Solution3 {
        public int[] smallestRange(List<List<Integer>> nums) {
            var pq = new PriorityQueue<int[]>(Comparator.comparingInt(a -> a[0]));
            int maxVal = Integer.MIN_VALUE, rangeStart = 0, rangeEnd = Integer.MAX_VALUE;

            for (int i = 0; i < nums.size(); i++) {
                pq.offer(new int[] { nums.get(i).get(0), i, 0 });
                maxVal = Math.max(maxVal, nums.get(i).get(0));
            }

            while (pq.size() == nums.size()) {
                int[] data = pq.poll();
                int minVal = data[0], row = data[1], col = data[2];

                if (maxVal - minVal < rangeEnd - rangeStart) {
                    rangeStart = minVal;
                    rangeEnd = maxVal;
                }

                if (col + 1 < nums.get(row).size()) {
                    int nextVal = nums.get(row).get(col + 1);
                    pq.offer(new int[] { nextVal, row, col + 1 });
                    maxVal = Math.max(maxVal, nextVal);
                }
            }

            return new int[] { rangeStart, rangeEnd };
        }
    }
}

File: src/leetcode_grind/day_712.rs
// https://leetcode.com/problems/maximum-number-of-moves-in-a-grid/description/?envType=daily-question&envId=2024-10-29
pub fn max_moves_i(grid: Vec<Vec<i32>>) -> i32 {
    let m = grid.len();
    let n = grid[0].len();

    let mut dp = vec![vec![0; n]; m];
    for i in 0..m {
        dp[i][0] = 1;
    }

    let mut max_moves = 0;
    for j in 1..n {
        for i in 0..m {
            if grid[i][j] > grid[i][j - 1] && dp[i][j - 1] > 0 {
                dp[i][j] = dp[i][j].max(dp[i][j - 1] + 1);
            }
            if i >= 1 && grid[i][j] > grid[i - 1][j - 1] && dp[i - 1][j - 1] > 0 {
                dp[i][j] = dp[i][j].max(dp[i - 1][j - 1] + 1);
            }
            if i + 1 < m && grid[i][j] > grid[i + 1][j - 1] && dp[i + 1][j - 1] > 0 {
                dp[i][j] = dp[i][j].max(dp[i + 1][j - 1] + 1);
            }
            max_moves = max_moves.max(dp[i][j] - 1);
        }
    }

    max_moves
}

pub fn max_moves_ii(grid: Vec<Vec<i32>>) -> i32 {
    let m = grid.len();
    let n = grid[0].len();

    let mut dp = vec![vec![0; 2]; m];
    for i in 0..m {
        dp[i][0] = 1;
    }
    let mut max_moves = 0;
    for j in 1..n {
        for i in 0..m {
            if grid[i][j] > grid[i][j - 1] && dp[i][0] > 0 {
                dp[i][1] = dp[i][1].max(dp[i][0] + 1);
            }
            if i >= 1 && grid[i][j] > grid[i - 1][j - 1] && dp[i - 1][0] > 0 {
                dp[i][1] = dp[i][1].max(dp[i - 1][0] + 1);
            }
            if i + 1 < m && grid[i][j] > grid[i + 1][j - 1] && dp[i + 1][0] > 0 {
                dp[i][1] = dp[i][1].max(dp[i + 1][0] + 1);
            }

            max_moves = max_moves.max(dp[i][1] - 1);
        }

        for k in 0..m {
            dp[k][0] = dp[k][1];
            dp[k][1] = 0;
        }
    }

    max_moves
}

// https://leetcode.com/problems/paint-house-ii/description/?envType=weekly-question&envId=2024-10-29
pub fn min_cost_ii(costs: Vec<Vec<i32>>) -> i32 {
    let (n, m) = (costs.len(), costs[0].len());
    let mut dp = vec![vec![1000_000_000; m]; n + 1];
    dp[0] = vec![0; m];
    for i in 1..=n {
        for j in 0..m {
            for k in 0..m {
                if j == k {
                    continue;
                }
                dp[i][j] = dp[i][j].min(costs[i - 1][j] + dp[i - 1][k])
            }
        }
    }
    dp[n].iter().copied().min().unwrap()
}

// https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description/
pub fn smallest_range(nums: Vec<Vec<i32>>) -> Vec<i32> {
    let k = nums.len();
    let mut indices = vec![0; k];
    let mut range = vec![0, i32::MAX];

    while true {
        let mut curr_min = i32::MAX;
        let mut curr_max = i32::MIN;
        let mut min_list_idx = 0;

        for i in 0..k {
            let mut curr_element = nums[i][indices[i]];

            if curr_element < curr_min {
                curr_min = curr_element;
                min_list_idx = i;
            }

            if curr_element > curr_max {
                curr_max = curr_element;
            }
        }

        if curr_max - curr_min < range[1] - range[0] {
            range[0] = curr_min;
            range[1] = curr_max;
        }

        indices[min_list_idx] += 1;
        if indices[min_list_idx] == nums[min_list_idx].len() {
            break;
        }
    }
    range
}
